### PR TITLE
refactor: collateral fix

### PR DESCRIPTION
**Pull request by toteki**
## Description

Collateral, when enabled, needs to be stored in the `x/leverage` module account instead of user wallets.

This PR changes the behavior of collateral as such, and as a consequence changes where various functions look for, send, or retrieve collateral.

closes: #286 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/keeper/keeper.go [hunk 1]
```diff
@@ -74,18 +74,27 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
...
 	}
 
+	// mint uToken
 	uTokens := sdk.NewCoins(uToken)
 	if err = k.bankKeeper.MintCoins(ctx, types.ModuleName, uTokens); err != nil {
 		return err
 	}
 
-	if err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, lenderAddr, uTokens); err != nil {
+	if k.GetCollateralSetting(ctx, lenderAddr, uToken.Denom) {
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper.go:89]: So we send `uTokens` to the module account if the `uToken.Denom` is NOT enabled as collateral?

File: x/leverage/keeper/keeper.go [hunk 2]
```diff
@@ -74,18 +74,27 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
...
+		// For uToken denoms enabled as collateral by this lender, the
+		// minted uTokens stay in the module account and the keeper tracks the amount
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper.go:91]:
```suggestion
		// minted uTokens stay in the module account and the keeper tracks the amount.
```

File: x/leverage/keeper/keeper.go (hunk 1 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go:89]: The opposite - collateral-enabled `uTokens` stay in the module (contributing to the lender's collateral amount), and collateral-disabled `uTokens` go in the lender's account.

Comment (alexanderbez) [reply x/leverage/keeper/keeper.go:89]: Is there a way we can refactor this so it's a bit more clear? In the case where it's disabled, we send to the module account and then back again costing the user more gas. Perhaps we move the sending to an if/else at the bottom or something.

Comment (toteki) [reply x/leverage/keeper/keeper.go:89]: I see the misunderstanding - the `SendCoinsFromAccountToModule(ctx, lenderAddr, types.ModuleName, loanTokens)` is base assets and the `SendCoinsFromModuleToAccount(ctx, types.ModuleName, lenderAddr, uTokens)` is uTokens.

So the transfers are:
- In all cases, `loanTokens` go from lender to module
- When collateral disabled, `uTokens` go from module to lender
- When collateral enabled, there is no second transfer

File: x/leverage/keeper/keeper.go [hunk 3]
```diff
@@ -96,42 +105,65 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
...
-	// send the uTokens from the lender to the module account
-	uTokens := sdk.NewCoins(uToken)
+	reservedAmount := k.GetReserveAmount(ctx, token.Denom)
+	availableAmount := k.bankKeeper.GetBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), token.Denom).Amount
+	if token.Amount.GT(availableAmount.Sub(reservedAmount)) {
+		return sdkerrors.Wrap(types.ErrLendingPoolInsufficient, token.String())
+	}
+
+	// Withdraw will first attempt to use any uTokens in the lender's wallet
+	fromWallet := sdk.MinInt(k.bankKeeper.GetBalance(ctx, lenderAddr, uToken.Denom).Amount, uToken.Amount)
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper.go:126]: consider a different variable name here 👍

File: x/leverage/keeper/keeper.go [hunk 4]
```diff
@@ -96,42 +105,65 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
...
+			// current borrowed value.
+			//
+			// ref: https://github.com/umee-network/umee/issues/213
+
+			// reduce the lender's collateral by fromCollateral
+			currentCollateral := k.GetCollateralAmount(ctx, lenderAddr, uToken.Denom)
+			if currentCollateral.Amount.LT(fromCollateral) {
+				return sdkerrors.Wrap(types.ErrInsufficientBalance, uToken.String())
+			}
+			newCollateral := sdk.NewCoin(uToken.Denom, currentCollateral.Amount.Sub(fromCollateral))
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper.go:143]:
```suggestion

			newCollateral := sdk.NewCoin(uToken.Denom, currentCollateral.Amount.Sub(fromCollateral))
```

File: x/leverage/keeper/keeper.go [hunk 5]
```diff
@@ -96,42 +105,65 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
...
+	// Any additional uTokens must come from the lender's collateral
+	fromCollateral := uToken.Amount.Sub(fromWallet)
+
+	if fromCollateral.IsPositive() {
+		if k.GetCollateralSetting(ctx, lenderAddr, uToken.Denom) {
+			// TODO: Calculate lender's borrow limit and current borrowed value, if any.
+			// Prevent withdrawing collateral when it would bring user borrow limit below
+			// current borrowed value.
+			//
+			// ref: https://github.com/umee-network/umee/issues/213
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/keeper.go:136]: Yeah, next PR will implement all borrow limit based restrictions

File: x/leverage/keeper/keeper.go (hunk 3 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go]: now `amountFromWallet` and `amountFromCollateral`

File: x/leverage/keeper/keeper.go [hunk 6]
```diff
@@ -96,42 +105,65 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
...
-	reservedAmount := k.GetReserveAmount(ctx, withdrawal.Denom)
-	availableAmount := k.bankKeeper.GetBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), withdrawal.Denom).Amount
-	if withdrawal.Amount.GT(availableAmount.Sub(reservedAmount)) {
-		return sdkerrors.Wrap(types.ErrLendingPoolInsufficient, withdrawal.String())
-	}
-
-	// send the uTokens from the lender to the module account
-	uTokens := sdk.NewCoins(uToken)
+	reservedAmount := k.GetReserveAmount(ctx, token.Denom)
+	availableAmount := k.bankKeeper.GetBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), token.Denom).Amount
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/keeper.go]: Done, it's in keeper.go for now

File: x/leverage/keeper/keeper.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply x/leverage/keeper/keeper.go:153]: ACK 👍

**Review by alexanderbez — APPROVED** [review]
🎉 🎉 🎉

**Pull request merged**